### DISCUSSION TITLE
I made a Node.js loader with Oxc and wasm-pack

### DISCUSSION THREAD

**Discussion by egasimus** [oxc-project/oxc#2105] [General]
Compiling TS on demand has been a major pain point for me. So when they dragged me kicking and screaming onto the TypeScript bandwagon, first thing I did was write a Node.js loader to make sure "build and run" stays one command and not two. (That, and the extensions.) :dizzy: 

That loader was **@hackbg/ganesha**, and today I upgraded it to use a WebAssembly build of Oxc, courtesy of [wasm-pack](https://rustwasm.github.io/docs/wasm-pack/) - as well as the stabilizing Node 20+ [Loader API](https://nodejs.org/api/module.html#customization-hooks).

If you're interested, the code is at https://github.com/hackbg/ganesha and the package at https://www.npmjs.com/package/@ganesha/oxc :elephant: 

Some more info: Ganesha v4 used `esbuild`, and was getting excruciatingly slow for large repos (like the one I'm wrangling currently). v4 had also accumulated a considerable amount of legacy cruft, which is not currently relevant to my work. Unlike v4, the Oxc-based **Ganesha v5** is short and sweet; haven't benchmarked yet, but so far v5 feels at least twice as fast. :runner: 

Plus, putting it together was a breeze! I'm honestly surprised at how smoothly it all worked out :rocket: 

The main feature it's currently missing is source map support, so that the line numbers in stack traces would make sense. Where should I look for source maps? (tbh for Node loaders I'd prefer to not use source maps, but instead replace all TS-specific tokens with whitespace, so that line/char numbers would match in the source and compiled code; source maps I'd rather use when bundling/minifying :thinking:)

**Comment by overlookmotel** [comment]
Very nice!

Out of interest, why did you use WASM, rather than Node-API (via [napi-rs](https://napi.rs/))? I imagine a NAPI-RS version would be a bit faster.

**Reply by egasimus** [reply]
Well, basically, I don't want to deal with native binaries :grin:

**Comment by overlookmotel** [comment]
I had a look at the code for your loader. You're right, it's surprisingly (and pleasantly) simple!

However, I've found what I think is a memory usage problem in your current implementation. Have opened an issue on the repo https://github.com/hackbg/ganesha/issues/23.

**Reply by Boshen** [reply]
Note to self: document this, also document errors emitted from the semantic builder.
